registry/storage/driver/s3-aws/s3.go

```diff
@@ -1401,12 +1343,8 @@ func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
 
 func (w *writer) Write(p []byte) (int, error) {
-	if w.closed {
-		return 0, fmt.Errorf("already closed")
-	} else if w.committed {
-		return 0, fmt.Errorf("already committed")
-	} else if w.cancelled {
-		return 0, fmt.Errorf("already cancelled")
+	if err := w.done(); err != nil {
+		return 0, err
 	}
 
 	// If the last written part is smaller than minChunkSize, we need to make a
```

Comment:
I'd really like to move the code in this condition (`if len(w.parts) > 0 && int(*w.parts[len(w.parts)-1].Size) < minChunkSize {`) for clarity into its own function. It would make the whole `Write` function much easier to grok, though I am afraid it would be a bit distracting for this change.